lib/WUI/espif.cpp

```diff
@@ -0,0 +1,538 @@
+
+    xSemaphoreTake(uart_write_mutex, portMAX_DELAY);
+    espif_transmit_data(INTRON, sizeof(INTRON));
+    uint8_t msg_type = MSG_PACKET;
+    espif_transmit_data(&msg_type, 1);
+    espif_transmit_data(&len, sizeof(len));
+    while (p != NULL) {
+        if (espif_transmit_data(p->payload, p->len) != ERR_OK) {
+            log_error(ESPIF, "Low level output packet failed");
+            xSemaphoreGive(uart_write_mutex);
+            return ERR_IF;
+        }
+        p = p->next;
+    }
+
+    xSemaphoreGive(uart_write_mutex);
+    return ERR_OK;
+}
+
+static err_t reset_and_wait() {
+    // Capture this task handle to manage wakeup from input thread
+    init_task_handle = xTaskGetCurrentTaskHandle();
+
+    // Reset device to receive MAC address
+    log_debug(ESPIF, "Resetting ESP to receive MAC address");
+    hard_reset_device();
+
+    // Wait for esp ready (receiveing device info)
+    if (!xTaskNotifyWait(0, 0, NULL, pdMS_TO_TICKS(3000))) {
+        log_error(ESPIF, "Timeout waiting for ESP bootup");
+        return ERR_IF;
+    }
+    log_info(ESPIF, "ESP up and running");
+    return ERR_OK;
+}
+
+/**
+ * @brief Initalize ESPIF network interface
+ *
+ * This initializes NET interface. Can be called multiple times if ESP modes are switched.
+ *
+ * @param netif Interface to initialize
+ * @return err_t Possible error encountered during initialization
+ */
+err_t espif_init(struct netif *netif) {
+    log_info(ESPIF, "LwIP init");
+
+    espif_reconfigure_uart(NIC_UART_BAUDRATE);
+    esp_operating_mode = ESPIF_RUNNING_MODE;
+
+    // Create mutex to protect UART writes
+    uart_write_mutex = xSemaphoreCreateMutex();
+    if (!uart_write_mutex) {
+        return ERR_IF;
+    }
+
+    // Create UART reader thread
+    uart_reader_id = osThreadCreate(osThread(UartBufferThread), netif);
+    if (uart_reader_id == NULL) {
+        log_error(ESPIF, "Failed to start UART reader");
+        return ERR_MEM;
+    }
+
+    // Initialize lwip netif
+    netif->name[0] = 'w';
+    netif->name[1] = 'l';
+    netif->output = etharp_output;
+#if LWIP_IPV6
+    netif->output_ip6 = ethip6_output;
+#endif
+    netif->linkoutput = low_level_output;
+
+    // LL init
+    netif->hwaddr_len = 0;
+    // TODO: This assumes LwIP can live with hwaddr not being set until ESP reports it
+    netif->mtu = 1500;
+    netif->flags = NETIF_FLAG_BROADCAST | NETIF_FLAG_ETHARP;
+
+    return reset_and_wait();
+}
+
+err_t espif_flash_initialize() {
+    // NOTE: There is no extra synchronization with reader thread. This assumes
+    // it is not a problem if reader thread reads some garbage until it notices
+    // operating mode change.
+    // NOTE: This holds the writer mutex only during this call. Holding this one
+    // all the time the ESP is being flashed might block LwIP thread and prevent
+    // ethernet from being serviced. Still, all the writers must have finished -
+    // this holds the lock and new writers will fail as mode is set to flashing.
```

Comment:
Let's hope you're right; I'll think about it more later… but I'm afraid there might be some race conditions.

Reply:
You are right, it is not 100% safe. I do not like UI running the flashing. When the UI menu is left (i.e. by timer) the flash process is stopped in middle. I think, in future, this should be refactored so that espif code would run the whole process while the UI would only report progress. That would allow for proper mutual exclusion of flashing and normal operation.

Reply:
Moving flashing into this thread sounds like the right solution, long term, yes. I just hope it'll be some manageable amount of work.